### ISSUE TITLE
Process#run broken with defunct child process with preview_mt

### ISSUE THREAD

**Issue by phil294** [crystal-lang/crystal#12392]
## Bug Report

Hi,
this seems very similar to #12241, however that issue is about the interpreter. This one is not, and the setup is different.

```
Crystal 1.5.0 (2022-07-21). LLVM: 14.0.6. Default target: x86_64-pc-linux-gnu
```
This source
```crystal
spawn do
	i = 0
	50_000_000_000_i64.times do
		i += 1
	end
end

150.times do
	stdout = IO::Memory.new
	puts "before run"
	Process.run("ls", output: stdout)
	puts "after run"
	stdout.close
	sleep 10.milliseconds
end
```
With `-Dpreview_mt`, in development (*without* `--release`), this is the output:
     ```
     before run
     after run
     before run
     ```
And then it freezes. That is, until the top thread finishes, after which the remaining processes get created properly.

Sleep in the code is not necessary, but increasing sleep also does not help.

This does not occur without the upper blocking thread (doesn't matter what it's doing, as long as it is blocking), and also not without the process `output` option.

Increasing `CRYSTAL_WORKERS` leads more iterations. In other words, every `Process.run` occupies a worker thread but does not release it again.

`ps` always shows a single `[ls] <defunct>`.

When you build with `--release` however, the blocking is gone and it executes as expected.

**Comment by straight-shoota** [comment]
When the `stdout` parameter is an `IO`, the process implementation creates a file system pipe, hands that to the forked process and then spawns a fiber to copy from the pipe into the IO.

That's the most likely reason for this to break.
I suspect one of the copy fibers gets scheduled on the same thread as the blocking work fiber. The `mt_preview` scheduler does not do work stealing, so it sits there until the blocking fiber finishes. And `Process.run` in the main fiber blocks until all copy fibers are completed.

For this reason, spawning a fiber with blocking code is not recommended. Fibers in Crystal are expected to be cooperative. The documentation requests to insert `Fiber.yield` into computation-heavy tasks (https://crystal-lang.org/api/1.5.0/Fiber.html#cooperative). A blocking C function cannot fulfill that.

We're currently missing a proper way to implement blocking function calls in a dedicated thread. This is discussed in #11778.

**Comment by ysbaddaden** [comment]
It can also block if the fiber that handles signals is on the same thread as the CPU-bound fiber, because `Process.run` waits for `SIGCHLD` to complete.

**Comment by zw963** [comment]
It's seem like following code work as expected on 1.17.1

```crystal
iso = Fiber::ExecutionContext::Isolated.new("worker") do
  i = 0_i64
  50_000_000_000_i64.times do
    i += 1
  end
end

10.times do
  stdout = IO::Memory.new
  puts "before run"
  Process.run("ls", output: stdout)
  puts "after run"
  sleep 10.milliseconds
end

iso.wait
```

You need run it with: `-Dpreview_mt -Dexecution_context`

**Comment by straight-shoota** [comment]
Yes, this kind of issue should be expected to be resolved with execution context. That's still a preview feature for now, but we can probably close this issue.
